app/controllers/api/v0/events_controller.rb

```diff
@@ -16,23 +16,59 @@ def create
     render nothing: true, status: 201
   end
 
-  private
+  class KafkaData
+    def initialize(api_data:, controller:)
+      @data = api_data.to_hash
+      @controller = controller
 
-  def convert_api_data_to_kafka_data(api_data:)
-    api_data.symbolize_keys.tap do |data|
-      if data[:type].include?('org.openstax.ec.started_session')
-        data[:ip_address] = request.remote_ip
-        data[:user_agent] = request.headers['User-Agent']
-      end
+      translate_started_session
+      translate_source_uri
+      translate_uuids
+      fixup_timestamps
+    end
+
+    def to_hash
+      data
+    end
+
+    protected
+
+    attr_reader :data, :controller
+
+    def translate_started_session
+      # Capture request-level data for session starts
+      return unless data[:type].include?('org.openstax.ec.started_session')
+      data[:ip_address] = controller.request.remote_ip
+      data[:user_agent] = request.headers['User-Agent']
+    end
 
+    def translate_source_uri
+      # Split source_uri into parts
+      # URI translates anything other than nil, and the swagger schema requires
+      # source_uri to be non-nil
+      # TODO parse query params into a hash of some sort?
```

Comment:
remove TODO?

Reply:
Got an opinion? Should we parse them, or store as a typical URL-string-mess? Do we have query params as part of our usual sites? Enquiring minds want to know!

Reply:
Umm, talked w/ Tom, we decided to parse them, and I did. They actually come out nicely in the avro schema, looks like you would even be able to filter on them when fetching eventts from kafka.